File: app/src/main/java/yagoc/ComputerPlayer.java
package yagoc;

import java.io.Serializable;
import java.util.LinkedList;
import java.util.function.BiFunction;

class ComputerPlayer implements Player, Serializable {
    final protected Logger logger;
    final protected SetType set;
    final private String name;
    final private int level;
    int processedMoves;
    private final BiFunction<Board, SetType, Integer> strategy;

    ComputerPlayer(String name, SetType set, Logger logger, int level, BiFunction<Board, SetType, Integer> strategy) {
        this.name = name;
        this.set = set;
        this.logger = logger;
        this.level = level;
        this.strategy = strategy;
    }

    @Override
    public PlayerType type() {
        return PlayerType.computer;
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public String toString() {
        return set + "\t" + name() + "\t" + type() + "\t" + level;
    }

    public Move move(Board board) {
        processedMoves = 0;
        MoveValue moveValue = alphaBeta(level, board, Integer.MIN_VALUE, Integer.MAX_VALUE);
        logger.info("alpha-beta: moves processed = " + processedMoves + ",  minimax = " + moveValue.value);

        return moveValue.move;
    }

    MoveValue alphaBeta(int depth, Board board, int alfa, int beta) {
        if (depth == 0) {
            return leafMoveValue(board);
        } else if ((level - depth) % 2 == 0) { // maximizing player = current player (as depth = level)
            return alphaBetaMax(depth, board, alfa, beta);
        } else {
            return alphaBetaMin(depth, board, alfa, beta);
        }
    }

    private MoveValue alphaBetaMin(int depth, Board board, int alpha, int beta) {
        LinkedList<Move> moves = new LinkedList<>(board.generateMoves());

        // checkmate
        if (moves.size() == 0) {
            return new MoveValue(Integer.MAX_VALUE - (level - depth + 1));
        }

        MoveValue betaMoveValue;
        Move move;
        MoveResult moveResult;
        MoveValue moveValue = new MoveValue(beta);

        do {
            processedMoves++;
            move = moves.removeFirst();
            moveResult = board.play(move);

            // beta = min[beta, AlphaBeta(N_k,alpha,beta)]
            betaMoveValue = alphaBeta(depth - 1, board, alpha, moveValue.value);
            board.undo(moveResult);

            if (betaMoveValue.value < moveValue.value) {
                moveValue = new MoveValue(move, betaMoveValue.value); // better
            }

            // beta cutoff
            if (alpha >= moveValue.value) {
                return new MoveValue(moveValue.move, alpha);
            }
        } while (moves.size() != 0);

        //si pierde MIN en 2 o más movimientos dejamos que juegue hasta entonces
        if (moveValue.move == null) {
            moveValue = new MoveValue(move, moveValue.value);
        }

        return moveValue;
    }

    private MoveValue alphaBetaMax(int depth, Board board, int alpha, int beta) {
        LinkedList<Move> moves = new LinkedList<>(board.generateMoves());

        // checkmate
        if (moves.size() == 0) {
            return new MoveValue(Integer.MIN_VALUE + (level - depth + 1));
        }

        MoveValue alphaMoveValue;
        Move move;
        MoveResult moveResult;
        MoveValue moveValue = new MoveValue(alpha);

        do {
            processedMoves++;
            move = moves.removeFirst();
            moveResult = board.play(move);

            // alpha = max[alpha, AlphaBeta(N_k,alpha,beta)
            alphaMoveValue = alphaBeta(depth - 1, board, moveValue.value, beta);
            board.undo(moveResult);

            if (alphaMoveValue.value > moveValue.value) {
                moveValue = new MoveValue(move, alphaMoveValue.value); // better
            }

            // alpha cutoff
            if (moveValue.value >= beta) {
                return new MoveValue(moveValue.move, beta);
            }
        } while (moves.size() != 0);

        //si pierde MAX en 2 o más movimientos dejamos que juegue hasta entonces
        if (moveValue.move == null) {
            moveValue = new MoveValue(move, moveValue.value);
        }

        return moveValue;
    }

    protected MoveValue leafMoveValue(Board board) {
        return new MoveValue(strategy.apply(board, set));
    }
}

File: yagoc-desktop/src/main/kotlin/jenm/yagoc/players/ComputerPlayer.kt
package jenm.yagoc.players

import jenm.yagoc.Yagoc
import jenm.yagoc.board.BoardRules.generateMoves
import jenm.yagoc.board.BoardView
import jenm.yagoc.board.Move
import jenm.yagoc.board.a1Square
import jenm.yagoc.pieces.PieceColor
import jenm.yagoc.pieces.blackPawn
import java.io.Serializable
import java.util.concurrent.atomic.AtomicInteger

class ComputerPlayer(
    name: String,
    pieceColor: PieceColor,
    private val level: Int,
    private val strategy: PlayerStrategy
) : Player(name, pieceColor, PlayerType.Computer), Serializable {
    private val noMoves = Move(blackPawn, a1Square, a1Square) //TODO find another way to represent this scenario

    override fun toString() = pieceColor.toString() + "\t" + type + "\t" + level

    override fun move(board: BoardView): Move {
        val moveCounter = AtomicInteger(0)
        val moveValue = alphaBeta(level, board, Int.MIN_VALUE, Int.MAX_VALUE, moveCounter)
        Yagoc.logger.info("processed = " + moveCounter + ", minimax = " + moveValue.value)
        return moveValue.move
    }

    fun alphaBeta(depth: Int, board: BoardView, alfa: Int, beta: Int, moveCounter: AtomicInteger): MoveValue = when {
        depth == 0 -> leafValue(board)
        (level - depth) % 2 == 0 -> // maximizing player = current player (as depth = level)
            alphaBetaMax(depth, board, alfa, beta, moveCounter)
        else -> alphaBetaMin(depth, board, alfa, beta, moveCounter)
    }

    private fun alphaBetaMin(
        depth: Int,
        board: BoardView,
        alpha: Int,
        beta: Int,
        processedMoves: AtomicInteger
    ): MoveValue {
        val moves = generateMoves(board)

        // checkmate
        if (moves.isEmpty()) {
            return MoveValue(noMoves, Int.MAX_VALUE - (level - depth + 1))
        }
        var betaMoveValue: MoveValue
        var moveValue = MoveValue(moves.first(), beta)
        for (move in moves) {
            processedMoves.getAndIncrement()

            // beta = min[beta, AlphaBeta(N_k,alpha,beta)]
            val v = moveValue.value
            betaMoveValue = board.playAndUndo(move) { alphaBeta(depth - 1, board, alpha, v, processedMoves) }
            if (betaMoveValue.value < moveValue.value) {
                moveValue = MoveValue(move, betaMoveValue.value) // better
            }

            // beta cutoff
            if (alpha >= moveValue.value) {
                return MoveValue(moveValue.move, alpha)
            }
        }

        return moveValue
    }

    private fun leafValue(board: BoardView) = MoveValue(noMoves, strategy(board, pieceColor))

    private fun alphaBetaMax(
        depth: Int,
        board: BoardView,
        alpha: Int,
        beta: Int,
        processedMoves: AtomicInteger
    ): MoveValue {
        val moves = generateMoves(board)

        // checkmate
        if (moves.isEmpty()) {
            return MoveValue(noMoves, Int.MIN_VALUE + (level - depth + 1))
        }
        var alphaMoveValue: MoveValue
        var moveValue = MoveValue(moves.first(), alpha)
        for (move in moves) {
            processedMoves.getAndIncrement()

            // alpha = max[alpha, AlphaBeta(N_k,alpha,beta)
            val v = moveValue.value
            alphaMoveValue = board.playAndUndo(move) { alphaBeta(depth - 1, board, v, beta, processedMoves) }
            if (alphaMoveValue.value > moveValue.value) {
                moveValue = MoveValue(move, alphaMoveValue.value) // better
            }

            // alpha cutoff
            if (moveValue.value >= beta) {
                return MoveValue(moveValue.move, beta)
            }
        }

        return moveValue
    }
}